controllers/metallb_controller.go

```
@@ -269,3 +241,133 @@ func valueWithDefault(name string, def int) (int, error) {
+			"create": false,
+			"name":   "controller",
+		},
+		"logLevel": logLevel,
+	}
+}
+
+func patchSpeakerValues(chartValues map[string]interface{}, config *metallbv1beta1.MetalLB) error {
+	logLevel := metallbv1beta1.LogLevelInfo
+	if config.Spec.LogLevel != "" {
+		logLevel = config.Spec.LogLevel
+	}
+	speakerImage := os.Getenv("SPEAKER_IMAGE")
+	speakerRepo, speakerTag := getImageNameTag(speakerImage)
+	if speakerRepo == "" {
+		speakerRepo = defaultSpeakerImageRepo
+	}
+	frrEnabled := false
+	var frrRepo, frrTag string
+	if os.Getenv("METALLB_BGP_TYPE") == bgpFrr {
+		frrEnabled = true
+		frrRepo, frrTag = getImageNameTag(os.Getenv("FRR_IMAGE"))
+		if frrRepo == "" {
+			frrRepo = defaultFRRImageRepo
+			frrTag = defaultFRRImageTag
+		}
+	}
+	mlBindPort, err := valueWithDefault("MEMBER_LIST_BIND_PORT", 7946)
+	if err != nil {
+		return err
+	}
+	frrMetricsPort, err := valueWithDefault("FRR_METRICS_PORT", 7473)
+	if err != nil {
+		return err
+	}
+	chartValues["speaker"] = map[string]interface{}{
+		"image": map[string]interface{}{
+			"repository": speakerRepo,
+			"tag":        speakerTag,
+		},
+		"serviceAccount": map[string]interface{}{
+			"create": false,
+			"name":   "speaker",
+		},
+		"frr": map[string]interface{}{
+			"enabled": frrEnabled,
+			"image": map[string]interface{}{
+				"repository": frrRepo,
+				"tag":        frrTag,
+			},
+			"metricsPort": frrMetricsPort,
+		},
+		"memberlist": map[string]interface{}{
+			"enabled":    true,
+			"mlBindPort": mlBindPort,
+		},
+		"logLevel": logLevel,
+	}
+	if config.Spec.SpeakerNodeSelector != nil {
+		chartValues["speaker"].(map[string]interface{})["nodeSelector"] = config.Spec.SpeakerNodeSelector
+	}
+	if config.Spec.SpeakerTolerations != nil {
+		chartValues["speaker"].(map[string]interface{})["tolerations"] = config.Spec.SpeakerTolerations
+	}
+	return nil
+}
+
+func getImageNameTag(envString string) (string, string) {
+	img := strings.Split(envString, ":")
+	if len(img) == 0 {
+		return "", ""
```

I'd return an error here